### PR TITLE
Adding a Webhook endpoint to instantly sync Slack messages

### PR DESCRIPTION
**Pull request by lasryaric**
Adding support for Webhooks.

- Add support for `/webhooks/:secret/slack`:
  - simply dispatch temporal workflows to sync a message with its context, using the same code as the full sync workflow.
- The auth middleware now check for a secret token in `/webhooks/:webhook_secret/` on `/webhooks` endpoints.
- Fixed a few missing await + added an eslint rule to prevent that mistake from happening again.

### CODE REVIEW

File: connectors/.env.example [hunk 1]
```diff
@@ -1,4 +1,6 @@
 CONNECTORS_DATABASE_URI=CONNECTORS_DATABASE_URI=postgres://postgres:mysecretpassword@localhost:5432/postgres
 NANGO_SLACK_CONNECTOR_ID=nango_slack_integration_id
 NANGO_SECRET_KEY=sk...
-DUST_CONNECTORS_SECRET=L3AhrnR3YuYEx6otwUnbzit29C1hTYnp
\ No newline at end of file
+DUST_CONNECTORS_SECRET=L3AhrnR3YuYEx6otwUnbzit29C1hTYnp
+DUST_CONNECTORS_WEBHOOKS_SECRET=webhooks_secret
```
Comment (spolu) [inline connectors/.env.example:5]: How is `DUST_CONNECTORS_WEBHOOKS_SECRET` set up with Slack? How is the webhook URL set up in Slack?

Please add a comment that explains how everything works or in the initial design doc or a runbook, really up to you.

File: connectors/src/api_server.ts [hunk 2]
```diff
@@ -30,6 +31,8 @@ export function startServer(port: number) {
   );
   app.post("/connectors/sync/:connector_id", syncConnectorAPIHandler);
 
+  app.post("/webhooks/:webhook_secret/slack", webhookSlackAPIHandler);
```
Comment (spolu) [inline connectors/src/api_server.ts:34]: How are the webhooks going through `DUST_CONNECTORS_SECRET` ?

File: connectors/src/connectors/slack/index.ts [hunk 3]
```diff
@@ -68,7 +68,7 @@ export async function createSlackConnector(
         { transaction: t }
       );
 
-      launchSlackSyncWorkflow(connector.id.toString());
+      await launchSlackSyncWorkflow(connector.id.toString());
```
Comment (spolu) [inline connectors/src/connectors/slack/index.ts:71]: Please handle in a separate PR along with the implementation of deleteConnector.

File: connectors/.eslintrc.json [hunk 4]
```diff
@@ -38,5 +39,8 @@
   "env": {
     "node": true,
     "es6": true
+  },
```
Comment (spolu) [inline connectors/.eslintrc.json:42]: Please handle in a separate PR along with implementation of deleteConnector as discussed on Slack

**Review by spolu — CHANGES_REQUESTED** [review]
Initial quesitons + please split previous PR follow-ups in a separate PR

File: connectors/src/api_server.ts (hunk 2 above)
Comment (spolu) [reply connectors/src/api_server.ts:34]: Ignore comment. Saw the updated auth code.

File: connectors/src/connectors/slack/temporal/activities.ts [hunk 5]
```diff
@@ -159,6 +175,7 @@ export async function syncNonThreaded(
       messages.push(message);
     }
   }
+  messages.reverse();
```
Comment (spolu) [inline connectors/src/connectors/slack/temporal/activities.ts:178]: This seems to impact the initial sync. Was it wrong before?

File: connectors/.env.example (hunk 1 above)
Comment (lasryaric) [reply connectors/.env.example:5]: Started the beginning of a slack runbook:
https://www.notion.so/dust-tt/d151a5226dba4f8fbe094365da5f00d4?v=c3d749040f7143f7aa61b0d29ba53a83&p=e59f50c458a64a9ea64f4fd17dacaffb&pm=s

File: connectors/src/connectors/slack/temporal/activities.ts (hunk 5 above)
Comment (lasryaric) [reply connectors/src/connectors/slack/temporal/activities.ts:178]: yes it was. Messages were reversed for non threaded messages.
This is because we added the time bounds, so the order is automatically reversed by the API :/

File: connectors/.env.example (hunk 1 above)
Comment (spolu) [reply connectors/.env.example:5]: Great!

File: connectors/src/connectors/slack/temporal/client.ts [hunk 6]
```diff
@@ -55,17 +59,22 @@ export async function launchSlackSyncOneThreadWorkflow(
...
-        connectorId,
-        dataSourceConfig,
-        nangoConnectionId,
-        channelId,
-        threadTs,
-      ],
-      taskQueue: "slack-queue",
-      workflowId: workflowId,
-    });
+    const handle = await client.workflow.signalWithStart(
```
Comment (lasryaric) [inline connectors/src/connectors/slack/temporal/client.ts:62]: `signalWithStart()` will make sure a new workflow is started or an existing one will get the signal.

File: connectors/src/connectors/slack/temporal/workflows.ts [hunk 7]
```diff
@@ -130,26 +135,44 @@ export async function syncOneChannel(
...
   threadTs: string
 ) {
-  const slackAccessToken = await getAccessToken(nangoConnectionId);
-  const channel = await getChannel(slackAccessToken, channelId);
-  if (!channel.name) {
-    throw new Error(`Could not find channel name for channel ${channelId}`);
+  let signaled = false;
+  let debounceCount = 0;
+
+  setHandler(newWebhookSignal, () => {
```
Comment (lasryaric) [inline connectors/src/connectors/slack/temporal/workflows.ts:148]: the magic is happening here.

File: connectors/src/connectors/slack/temporal/workflows.ts [hunk 8]
```diff
@@ -126,3 +135,91 @@ export async function workspaceSyncOneChannel(
...
+    await syncThread(
+      dataSourceConfig,
+      slackAccessToken,
+      channelId,
+      channel.name,
+      threadTs,
+      connectorId
+    );
+  }
+  // /!\ Any signal received outside of the while loop will be lost, so don't make any async
```
Comment (spolu) [inline connectors/src/connectors/slack/temporal/workflows.ts:177]: Your comment hint that this could happen in the middle of the two async calls above? Is that the case?

File: connectors/src/connectors/slack/temporal/workflows.ts [hunk 9]
```diff
@@ -126,3 +135,91 @@ export async function workspaceSyncOneChannel(
...
+      slackAccessToken,
+      dataSourceConfig,
+      channelId,
+      channel.name,
+      startTsMs,
+      endTsMs,
+      connectorId
+    );
+  }
+  // /!\ Any signal received outside of the while loop will be lost, so don't make any async
```
Comment (spolu) [inline connectors/src/connectors/slack/temporal/workflows.ts:223]: Same comment here?

File: connectors/src/middleware/auth.ts [hunk 10]
```diff
@@ -42,3 +56,37 @@ export const authMiddleware = (
...
 };
+
+const _authMiddlewareWebhooks = (
+  req: Request,
+  res: Response<ConnectorsAPIErrorResponse>,
+  next: NextFunction
+) => {
+  if (req.path.startsWith("/webhooks")) {
+    const parts = req.path.split("/");
+    const indexOfWebhooks = parts.indexOf("webhooks");
```
Comment (spolu) [inline connectors/src/middleware/auth.ts:67]: No way to use path templates here?

Comment (spolu) [reply connectors/src/middleware/auth.ts:67]: Let’s just do “includes” no?

**Review by spolu — COMMENTED** [review]
Left a few questions on things that seem still a bit unclear to me 🙏

File: connectors/src/connectors/slack/temporal/workflows.ts (hunk 8 above)
Comment (spolu) [reply connectors/src/connectors/slack/temporal/workflows.ts:177]: Ok now I understand. Was thinking the async part was after the loop 👍

File: connectors/src/connectors/slack/temporal/workflows.ts (hunk 9 above)
Comment (spolu) [reply connectors/src/connectors/slack/temporal/workflows.ts:223]: Same as above, now understood 👍

File: connectors/src/middleware/auth.ts (hunk 10 above)
Comment (lasryaric) [reply connectors/src/middleware/auth.ts:67]: Yes that's a lot simpler. I will do that.

File: connectors/src/api/webhooks/webhookSlack.ts [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+  challenge?: string;
+  team_id?: string;
+  event?: {
+    channel?: string;
+    ts?: string; // slack message id
+    thread_ts?: string; // slack thread id
+    type?: string; // event type (eg: message)
+  };
+};
+
```
Comment (spolu) [inline connectors/src/api/webhooks/webhookSlack.ts:23]: Naming off?

File: connectors/src/api/webhooks/webhookSlack.ts [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+type ConnectorCreateResBody =
+  | { challenge: string }
+  | { message: string }
+  | ConnectorsAPIErrorResponse;
+
+const _webhookSlackAPIHandler = async (
+  req: Request<
+    { connector_provider: string },
+    ConnectorCreateResBody,
```
Comment (spolu) [inline connectors/src/api/webhooks/webhookSlack.ts:32]: Naming as well?

Comment (spolu) [reply connectors/src/api/webhooks/webhookSlack.ts:32]: Maybe it needed actually?

File: connectors/src/api/webhooks/webhookSlack.ts [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+};
+
+type ConnectorCreateResBody =
+  | { challenge: string }
+  | { message: string }
+  | ConnectorsAPIErrorResponse;
+
+const _webhookSlackAPIHandler = async (
+  req: Request<
+    { connector_provider: string },
```
Comment (spolu) [inline connectors/src/api/webhooks/webhookSlack.ts:31]: Unrelated?

**Review by spolu — CHANGES_REQUESTED** [review]
Few small nits left 👍

File: connectors/src/api/webhooks/webhookSlack.ts [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+  };
+};
+
+type SlackWebhookResBody =
+  | { challenge: string }
+  | null
+  | ConnectorsAPIErrorResponse;
+
+const _webhookSlackAPIHandler = async (
+  req: Request<{}, SlackWebhookResBody, SlackWebhookReqBody>,
```
Comment (spolu) [inline connectors/src/api/webhooks/webhookSlack.ts:30]: SlackWebhookResBody in the request?

**Review by spolu — COMMENTED** [review]
Left one comment, + linter/type-system is not happy

File: connectors/src/api/webhooks/webhookSlack.ts (hunk 14 above)
Comment (lasryaric) [reply connectors/src/api/webhooks/webhookSlack.ts:30]: yes, that's how it's defined:
```
// connectors/node_modules/@types/express/index.d.ts
    interface Request<
        P = core.ParamsDictionary,
        ResBody = any,
        ReqBody = any,
        ReqQuery = core.Query,
        Locals extends Record<string, any> = Record<string, any>
```

We do that in the other endpoints.

**Review by spolu — APPROVED** [review]
LGTM
